common/src/main/java/org/conscrypt/ClientSessionContext.java

```diff
@@ -131,6 +147,42 @@ private NativeSslSession getSession(String host, int port) {
         return null;
     }
 
+    private void putSession(HostAndPort key, NativeSslSession session) {
+        synchronized (sessionsByHostAndPort) {
+            List<NativeSslSession> sessions = sessionsByHostAndPort.get(key);
+            if (sessions == null) {
+                sessions = new ArrayList<NativeSslSession>();
+                sessionsByHostAndPort.put(key, sessions);
+            }
+            // To maintain the invariant that single- and multi-use sessions aren't
+            // mixed, check what the current list contains and remove those sessions if
+            // they're of the other type.
+            if (sessions.size() > 0) {
+                if (sessions.get(0).isSingleUse() != session.isSingleUse()) {
```

Comment:
Nit: You could just fold this into the parent with an &&

Reply:
Done.